File: report-generator/src/main/java/com/teamscale/report/testwise/jacoco/JaCoCoTestwiseReportGenerator.java
package com.teamscale.report.testwise.jacoco;

import com.teamscale.report.EDuplicateClassFileBehavior;
import com.teamscale.report.jacoco.dump.Dump;
import com.teamscale.report.testwise.jacoco.cache.CoverageGenerationException;
import com.teamscale.report.testwise.model.TestwiseCoverage;
import com.teamscale.report.testwise.model.builder.TestCoverageBuilder;
import com.teamscale.report.util.ClasspathWildcardIncludeFilter;
import com.teamscale.report.util.ILogger;
import org.jacoco.core.data.ExecutionData;
import org.jacoco.core.data.ExecutionDataReader;
import org.jacoco.core.data.ExecutionDataStore;
import org.jacoco.core.data.IExecutionDataVisitor;
import org.jacoco.core.data.ISessionInfoVisitor;
import org.jacoco.core.data.SessionInfo;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Collection;
import java.util.function.Consumer;

/**
 * Creates a XML report for an execution data store. The report is grouped by session.
 * <p>
 * The class files under test must be compiled with debug information otherwise no coverage will be collected.
 */
public class JaCoCoTestwiseReportGenerator {

	/** The execution data reader and converter. */
	private CachingExecutionDataReader executionDataReader;

	/** The filter for the analyzed class files. */
	private final ClasspathWildcardIncludeFilter locationIncludeFilter;

	/**
	 * Create a new generator with a collection of class directories.
	 *
	 * @param codeDirectoriesOrArchives Root directory that contains the projects class files.
	 * @param locationIncludeFilter     Filter for class files
	 * @param logger                    The logger
	 */
	public JaCoCoTestwiseReportGenerator(Collection<File> codeDirectoriesOrArchives,
										 ClasspathWildcardIncludeFilter locationIncludeFilter,
										 EDuplicateClassFileBehavior duplicateClassFileBehavior,
										 ILogger logger) throws CoverageGenerationException {
		this.locationIncludeFilter = locationIncludeFilter;
		this.executionDataReader = new CachingExecutionDataReader(logger);
		this.executionDataReader
				.analyzeClassDirs(codeDirectoriesOrArchives, locationIncludeFilter, duplicateClassFileBehavior);
	}

	/** Converts the given dumps to a report. */
	public TestwiseCoverage convert(File executionDataFile) throws IOException {
		TestwiseCoverage testwiseCoverage = new TestwiseCoverage();
		CachingExecutionDataReader.DumpConsumer dumpConsumer = executionDataReader
				.buildCoverageConsumer(locationIncludeFilter, testwiseCoverage::add);
		readAndConsumeDumps(executionDataFile, dumpConsumer);
		return testwiseCoverage;
	}

	/** Converts the given dumps to a report. */
	public void convertAndConsume(File executionDataFile, Consumer<TestCoverageBuilder> consumer) throws IOException {
		CachingExecutionDataReader.DumpConsumer dumpConsumer = executionDataReader
				.buildCoverageConsumer(locationIncludeFilter, consumer);
		readAndConsumeDumps(executionDataFile, dumpConsumer);
	}

	/** Reads the dumps from the given *.exec file. */
	private void readAndConsumeDumps(File executionDataFile, Consumer<Dump> dumpConsumer) throws IOException {
		try (InputStream input = new BufferedInputStream(new FileInputStream(executionDataFile))) {
			ExecutionDataReader executionDataReader = new ExecutionDataReader(input);
			DumpCallback dumpCallback = new DumpCallback(dumpConsumer);
			executionDataReader.setExecutionDataVisitor(dumpCallback);
			executionDataReader.setSessionInfoVisitor(dumpCallback);
			executionDataReader.read();
			// Ensure that the last read dump is also consumed
			dumpCallback.processDump();
		}
	}

	/** Collects execution information per session and passes it to the consumer . */
	private static class DumpCallback implements IExecutionDataVisitor, ISessionInfoVisitor {

		/** The dump that is currently being read. */
		private Dump currentDump = null;

		/** The store to which coverage is currently written to. */
		private ExecutionDataStore store;

		/** The consumer to pass {@link Dump}s to. */
		private Consumer<Dump> dumpConsumer;

		private DumpCallback(Consumer<Dump> dumpConsumer) {
			this.dumpConsumer = dumpConsumer;
		}

		@Override
		public void visitSessionInfo(SessionInfo info) {
			processDump();
			store = new ExecutionDataStore();
			currentDump = new Dump(info, store);
		}

		@Override
		public void visitClassExecution(ExecutionData data) {
			store.put(data);
		}

		private void processDump() {
			if (currentDump != null) {
				dumpConsumer.accept(currentDump);
				currentDump = null;
			}
		}
	}
}

File: report-generator/src/main/kotlin/com/teamscale/report/testwise/jacoco/JaCoCoTestwiseReportGenerator.kt
package com.teamscale.report.testwise.jacoco

import com.teamscale.report.EDuplicateClassFileBehavior
import com.teamscale.report.jacoco.dump.Dump
import com.teamscale.report.testwise.jacoco.CachingExecutionDataReader.*
import com.teamscale.report.testwise.jacoco.cache.CoverageGenerationException
import com.teamscale.report.testwise.model.TestwiseCoverage
import com.teamscale.report.testwise.model.builder.TestCoverageBuilder
import com.teamscale.report.util.ClasspathWildcardIncludeFilter
import com.teamscale.report.util.ILogger
import org.jacoco.core.data.ExecutionData
import org.jacoco.core.data.ExecutionDataReader
import org.jacoco.core.data.ExecutionDataStore
import org.jacoco.core.data.IExecutionDataVisitor
import org.jacoco.core.data.ISessionInfoVisitor
import org.jacoco.core.data.SessionInfo
import java.io.BufferedInputStream
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.util.function.Consumer

/**
 * Creates an XML report for an execution data store. The report is grouped by session.
 *
 * The class files under test must be compiled with debug information otherwise no coverage will be collected.
 */
open class JaCoCoTestwiseReportGenerator(
	codeDirectoriesOrArchives: Collection<File>,
	private val locationIncludeFilter: ClasspathWildcardIncludeFilter,
	duplicateClassFileBehavior: EDuplicateClassFileBehavior,
	logger: ILogger
) {
	/** The execution data reader and converter. */
	private val executionDataReader = CachingExecutionDataReader(
		logger, codeDirectoriesOrArchives, locationIncludeFilter, duplicateClassFileBehavior
	)

	init {
		// This has to be unsafe as mockito does not support mocking final classes
		updateClassDirCache()
	}

	/** Updates the probe cache of the [ExecutionDataReader]. */
	open fun updateClassDirCache() {
		executionDataReader.analyzeClassDirs()
	}

	/** Converts the given dumps to a report. */
	@Throws(IOException::class, CoverageGenerationException::class)
	open fun convert(executionDataFile: File): TestwiseCoverage {
		val testwiseCoverage = TestwiseCoverage()
		val dumpConsumer = executionDataReader.buildCoverageConsumer(locationIncludeFilter, testwiseCoverage::add)
		readAndConsumeDumps(executionDataFile, dumpConsumer)
		return testwiseCoverage
	}

	/** Converts the given dump to a report. */
	@Throws(CoverageGenerationException::class)
	open fun convert(dump: Dump): TestCoverageBuilder? {
		val testCoverageBuilders = mutableListOf<TestCoverageBuilder>()
		executionDataReader
			.buildCoverageConsumer(locationIncludeFilter, testCoverageBuilders::add)
			.accept(dump)
		return testCoverageBuilders.singleOrNull()
	}

	/** Converts the given dumps to a report. */
	@Throws(IOException::class)
	open fun convertAndConsume(executionDataFile: File, consumer: Consumer<TestCoverageBuilder>) {
		val dumpConsumer = executionDataReader.buildCoverageConsumer(locationIncludeFilter, consumer)
		readAndConsumeDumps(executionDataFile, dumpConsumer)
	}

	/** Reads the dumps from the given *.exec file. */
	@Throws(IOException::class)
	private fun readAndConsumeDumps(executionDataFile: File, dumpConsumer: DumpConsumer) {
		BufferedInputStream(FileInputStream(executionDataFile)).use { input ->
			ExecutionDataReader(input).apply {
				val dumpCallback = DumpCallback(dumpConsumer)
				setExecutionDataVisitor(dumpCallback)
				setSessionInfoVisitor(dumpCallback)
				read()
				dumpCallback.processDump()
			}
		}
	}

	/** Collects execution information per session and passes it to the consumer . */
	private class DumpCallback(private val consumer: DumpConsumer) : IExecutionDataVisitor, ISessionInfoVisitor {
		/** The dump that is currently being read. */
		private var currentDump: Dump? = null

		/** The store to which coverage is currently written to. */
		private var store: ExecutionDataStore? = null

		override fun visitSessionInfo(info: SessionInfo) {
			processDump()
			ExecutionDataStore().let {
				currentDump = Dump(info, it)
				store = it
			}
		}

		override fun visitClassExecution(data: ExecutionData) {
			store?.put(data)
		}

		fun processDump() {
			currentDump?.let {
				consumer.accept(it)
				currentDump = null
			}
		}
	}
}